actions/archive2disk/v1/pkg/archive/utils.go

```diff
@@ -0,0 +1,107 @@
+package archive
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"io"
+	"os"
+	"path/filepath"
+	// See https://github.com/opencontainers/go-digest#usage
+	_ "crypto/sha256"
+	_ "crypto/sha512"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+func extractTarDirectory(root string, checksum string, r io.Reader) error {
```

Comment:
Instead of passing the checksum to `extractTarDirectory` why not throw the checksum validation into an implementer of `io.Reader`. It would then be doing checksum readying/streaming verification and pass it through to the next stage of the pipeline and it doesn't care if its to gz or tar.

Reply:
@mmlb  I like this idea.  However, I'm not certain how to implement it.  I'm more of a novice/intermediate Go developer.  Also per an earlier comment.  Keeping things fast, is there a way to perform the checksum before extracting the contents.  That way if the checksum fails, it just fails w/o any extracted files.  I did think about this but thought the process would involve reading the file from the source twice twice and hence reduce the speed of the process.  Or if the file is downloaded and saved, where would it get saved.  At the mount location?